### ISSUE TITLE
Revise duk_get_prop() and variant return value

### ISSUE THREAD

**Issue by svaarala** [svaarala/duktape#1171]
Current behavior is to return 1 if property is "present" and 0 if it is "absent". This distinction is not available to Ecmascript code doing a `x = obj.prop`. Moreover, if the target is a Proxy the present/absent information is not even available, and `duk_get_prop()` and its variants currently always return 1 for proxies that implement the `get` trap.

See discussion in #1160, specifically https://github.com/svaarala/duktape/pull/1160#issuecomment-266239053.

Concrete alternatives are:
- Keep current behavior but document limitation for Proxies #1170
- Change return code to be 0 when the property read result is undefined (present or absent) and 1 otherwise.
- Remove the return value altogether and assume application call sites will use an explicit `duk_is_undefined()` or other type check on the result.

**Comment by fatcerberus** [comment]
For what it's worth, while I find the current built-in presence check useful *now*, when I first got started with Duktape I found the return value to be a bit confusing: For a while I kept making the mistake of assuming a false return value meant nothing was pushed.  So just something else to consider from an API design standpoint.

**Comment by svaarala** [comment]
Right, that would also be a logical API: a value would be pushed only when one was found. I did consider that but I find it error prone in the other direction: it's easy to ignore the return value when the property is just assumed to exist - which will then break down in mysterious ways if that assumption is off. Overall I find constant call stack behaviors less error prone than variable ones, and while it's always a trade-off between various goals, I usually prefer that over variable behaviors because calling code is sometimes sloppy with the return value interpretation.

**Comment by fatcerberus** [comment]
Yeah, I wasn't actually suggesting to push nothing, it was more a point in favor of removing the return value altogether :)

**Comment by svaarala** [comment]
Just to be clear, I don't mean "sloppy" in a bad way. Sometimes it's fine to say for example:

```c
int port;
duk_get_prop_string(ctx, -1, "port");
port = duk_to_int(ctx, -1);
duk_pop(ctx);
```

Not dealing with a missing property causes a zero port to be used, which can be handled by the same error handling path as when a port is not available.

**Comment by svaarala** [comment]
> Yeah, I wasn't actually suggesting to push nothing, it was more a point in favor of removing the return value altogether :)

Ah I see :)

**Comment by fatcerberus** [comment]
In light of how ES2015 and later treat `undefined` vs. `null` for optional function parameters, I think if the return value is kept, then IMO the "least surprising" behavior would be to return false if and only if the value retrieved is `undefined`.  Anything else (including a `null`) should return true.
